transformations/syntactically_diverse_paraphrase/sowreap/parse_utils.py

```diff
@@ -0,0 +1,186 @@
+    "PRN",
+    "QP",
+    "RRC",
+    "UCP",
+    "VP",
+    "WHADJP",
+    "WHAVP",
+    "WHNP",
+    "WHPP",
+    "WHADVP",
+    "X",
+    "ROOT",
+    "NP-TMP",
+    "PRT",
+]
+
+
+class Token(object):
+    def __init__(self, word, idx):
+        self.word = word
+        self.idx = idx
+
+    def __repr__(self):
+        return repr(self.word)
+
+
+class Node(object):
+    def __init__(self):
+        self.root = False
+        self.children = []
+        self.label = None
+        self.parent = None
+        self.phrase = ""
+        self.terminal = False
+        self.start_idx = 0
+        self.end_idx = 0
+
+
+class Sentence(object):
+    def __init__(self, parse_string):
+        self.num_nodes = 0
+        self.tree = self.get_tree(parse_string)
+        self.sent = self.tree.phrase
+
+        tokens = {}
+        for idx, tok in enumerate(self.tree.phrase.split(" ")):
+            tokens[idx] = Token(tok, idx)
+        self.tokens = tokens
+        self.num_tokens = len(tokens)
+
+    def get_tree(self, parse_string):
+        tree = self.contruct_tree_from_parse(parse_string, None)
+        tree = self.reduce_tree(tree)
+        phrase_whole = self.assign_phrases(tree, 0)
+        tree.phrase = phrase_whole
+        tree.start_idx = 0
+        tree.end_idx = len(phrase_whole.split(" "))
+        tree.parent_idx = -1
+        self.assign_ids(tree)
+        return tree
+
+    def assign_ids(self, tree):
+        tree.idx = self.num_nodes
+        self.num_nodes += 1
+
+        for child in tree.children:
+            child.parent_idx = tree.idx
+            self.assign_ids(child)
+
+    @staticmethod
+    def get_subtrees(parse_txt_partial):
+        parse_txt_partial = parse_txt_partial[1:-1]
+        if "(" in parse_txt_partial:
+            idx_first_lb = parse_txt_partial.index("(")
+            name_const = parse_txt_partial[:idx_first_lb].strip()
+            parse_txt_partial = parse_txt_partial[idx_first_lb:]
+            count = 0
+            partition_indices = []
+            for idx in range(len(parse_txt_partial)):
+                if parse_txt_partial[idx] == "(":
+                    count += 1
+                elif parse_txt_partial[idx] == ")":
+                    count -= 1
+                if count == 0:
+                    partition_indices.append(idx + 1)
+
+            partitions = []
+            part_idx_prev = 0
+            for i, part_idx in enumerate(partition_indices):
+                partitions.append(parse_txt_partial[part_idx_prev:part_idx])
+                part_idx_prev = part_idx
+        else:
+            temp = parse_txt_partial.split(" ")
+            name_const = temp[0]
+            partitions = [temp[1]]
+
+        return name_const, partitions
+
+    # constructs constituency tree from the parse string
+    @staticmethod
+    def contruct_tree_from_parse(parse_txt, node):
+
+        if parse_txt.startswith("("):
+            phrase_name, divisions = Sentence.get_subtrees(parse_txt)
```

Comment:
Is `phrase_name` the part of speech tag associated to the subpart of the sentence?
Overall the variable naming seems to have inconsistencies (sometimes it is `divisions`, sometimes it is `partitions`) and might be improved to ease the understanding of the code. Are those sub_trees as they are returned by `get_subtrees`? If yes, why not call the variable `sub_trees`?